From 423def578ea36b98de36be16df722a9e7b7b5767 Mon Sep 17 00:00:00 2001
From: Gurumoorthi Gnanasambandhan <gguru@qti.qualcomm.com>
Date: Wed, 24 Jun 2020 17:52:40 +0530
Subject: [PATCH 01/40] qca-hostapd: 1905 EAPOL support

MAP functionality needs additional 4 way keyhandshake to generate
MIC for 1905 packets. Add support for wpa state machine to
communicate with MAP application.

Change-Id: Icd9d1e2bd7c61bd86844fbd5a910633a138f105f
Signed-off-by: Gurumoorthi Gnanasambandhan <gguru@qti.qualcomm.com>
---
 hostapd/Makefile                  |   1 +
 hostapd/main.c                    |   4 +
 src/ap/wpa_auth.c                 |   2 +-
 src/ap/wpa_auth.h                 |   1 +
 src/drivers/eap1905.c             | 397 ++++++++++++++++++++++++++
 src/drivers/eap1905.h             |   7 +
 src/drivers/eap1905_sta.c         | 445 ++++++++++++++++++++++++++++++
 wpa_supplicant/Makefile           |   1 +
 wpa_supplicant/ctrl_iface_unix.c  |   3 +
 wpa_supplicant/main.c             |   6 +-
 wpa_supplicant/wpa_supplicant_i.h |   1 +
 11 files changed, 866 insertions(+), 2 deletions(-)
 create mode 100644 src/drivers/eap1905.c
 create mode 100644 src/drivers/eap1905.h
 create mode 100644 src/drivers/eap1905_sta.c

diff --git a/hostapd/Makefile b/hostapd/Makefile
index 09f873721b..3fd0ab049b 100644
--- a/hostapd/Makefile
+++ b/hostapd/Makefile
@@ -97,6 +97,7 @@ NEED_MD5=y
 NEED_SHA1=y
 
 OBJS += ../src/drivers/drivers.o
+OBJS += ../src/drivers/eap1905.o
 CFLAGS += -DHOSTAPD
 
 ifdef CONFIG_TAXONOMY
diff --git a/hostapd/main.c b/hostapd/main.c
index 58cc86a08c..82a5803e30 100644
--- a/hostapd/main.c
+++ b/hostapd/main.c
@@ -30,6 +30,7 @@
 #include "fst/fst.h"
 #include "config_file.h"
 #include "eap_register.h"
+#include "drivers/eap1905.h"
 #include "ctrl_iface.h"
 
 
@@ -1016,6 +1017,8 @@ int main(int argc, char *argv[])
 
 	hostapd_global_ctrl_iface_init(&interfaces);
 
+	eapol_1905_init();
+
 	if (hostapd_global_run(&interfaces, daemonize, pid_file)) {
 		wpa_printf(MSG_ERROR, "Failed to start eloop");
 		goto out;
@@ -1024,6 +1027,7 @@ int main(int argc, char *argv[])
 	ret = 0;
 
  out:
+	eapol_1905_deinit();
 	hostapd_global_ctrl_iface_deinit(&interfaces);
 	/* Deinitialize all interfaces */
 	for (i = 0; i < interfaces.count; i++) {
diff --git a/src/ap/wpa_auth.c b/src/ap/wpa_auth.c
index 2f90d77cc8..848f2181f8 100644
--- a/src/ap/wpa_auth.c
+++ b/src/ap/wpa_auth.c
@@ -770,7 +770,7 @@ struct wpa_authenticator * wpa_init(const u8 *addr,
 		return NULL;
 	}
 
-	wpa_auth->group = wpa_group_init(wpa_auth, 0, 1);
+	wpa_auth->group = wpa_group_init(wpa_auth, 0, !conf->no_group_delay);
 	if (!wpa_auth->group) {
 		os_free(wpa_auth->wpa_ie);
 		os_free(wpa_auth);
diff --git a/src/ap/wpa_auth.h b/src/ap/wpa_auth.h
index 103e8db76f..9af81dea99 100644
--- a/src/ap/wpa_auth.h
+++ b/src/ap/wpa_auth.h
@@ -268,6 +268,7 @@ struct wpa_auth_config {
 	unsigned int prot_range_neg:1;
 
 	int owe_ptk_workaround;
+	int no_group_delay;
 	u8 transition_disable;
 #ifdef CONFIG_DPP2
 	int dpp_pfs;
diff --git a/src/drivers/eap1905.c b/src/drivers/eap1905.c
new file mode 100644
index 0000000000..d1bec04c23
--- /dev/null
+++ b/src/drivers/eap1905.c
@@ -0,0 +1,397 @@
+
+/*
+ * Testing tool for EAPOL-Key Supplicant/Authenticator routines
+ * Copyright (c) 2006-2019, Jouni Malinen <j@w1.fi>
+ *
+ * This software may be distributed under the terms of the BSD license.
+ * See README for more details.
+ */
+
+#include "utils/includes.h"
+
+#include "utils/common.h"
+#include "utils/eloop.h"
+#include "rsn_supp/wpa.h"
+#include "ap/wpa_auth.h"
+
+#define RX_PORT 9900
+#define TX_PORT 9901
+#define MAX_AUTH 8
+
+struct wpa_1905 {
+	u8 auth_addr[ETH_ALEN];
+	u8 supp_addr[MAX_AUTH][ETH_ALEN];
+
+	struct wpa_auth_callbacks auth_cb;
+	struct wpa_authenticator *auth_group;
+	struct wpa_state_machine *auth[MAX_AUTH];
+
+	u8 supp_ie[80];
+	size_t supp_ie_len;
+	u8 psk[PMK_LEN];
+	u8 pmkid[PMKID_LEN];
+
+	int sta_count;
+};
+
+struct sec_config {
+	u8 auth_addr[ETH_ALEN];
+	u8 psk[PMK_LEN];
+	int pairwise_cipher;
+	int group_cipher;
+	int key_mgmt;
+	int wpa_proto;
+	int mgmt_group_cipher;
+	int ieee80211w;
+} STRUCT_PACKED;
+
+struct assoc {
+	u8 mac_addr[ETH_ALEN];
+	size_t supp_ie_len;
+	u8 supp_ie[80];
+	u8 pmkid[PMKID_LEN];
+} STRUCT_PACKED;
+
+struct key_params {
+	u8 mac_addr[ETH_ALEN];
+	int idx;
+	u8 key_data[32];
+	size_t key_len;
+} STRUCT_PACKED;
+
+enum events {
+	EVENT_AP_SEC_CONFIG,
+	EVENT_STA_SEC_CONFIG,
+	EVENT_1905_ASSOC,
+	EVENT_DISASSOC,
+	EVENT_1905_EAPOL,
+	EVENT_SETKEY,
+	EVENT_GROUP_KEY,
+};
+
+int eapol_1905_sock = -1;
+int eapol_1905_sock_tx = -1;
+
+static struct wpa_1905 wpa;
+struct wpa_auth_config conf;
+
+static void auth_logger(void *ctx, const u8 *addr, logger_level level,
+			const char *txt)
+{
+	if (addr)
+		wpa_printf(MSG_DEBUG, "AUTH: " MACSTR " - %s",
+			   MAC2STR(addr), txt);
+	else
+		wpa_printf(MSG_DEBUG, "AUTH: %s", txt);
+}
+
+static const u8 *auth_get_psk(void *ctx, const u8 *addr,
+			       const u8 *p2p_dev_addr, const u8 *prev_psk,
+			       size_t *psk_len, int *vlan_id)
+{
+	struct wpa_1905 *wpa = ctx;
+
+	wpa_printf(MSG_DEBUG, "AUTH: %s (addr=" MACSTR " prev_psk=%p)",
+		   __func__, MAC2STR(addr), prev_psk);
+	if (vlan_id)
+		*vlan_id = 0;
+	if (psk_len)
+		*psk_len = PMK_LEN;
+	if (prev_psk)
+		return NULL;
+	return wpa->psk;
+}
+
+static int auth_set_key(void *ctx, int vlan_id, enum wpa_alg alg,
+			const u8 *addr, int idx, u8 *key,
+			size_t key_len, enum key_flag key_flag)
+{
+	struct wpa_1905 *wpa = ctx;
+	char send_buffer[64] = {0} ;
+	struct key_params key_param;
+	os_memset(&key_param, 0, sizeof(key_param));
+
+	wpa_printf(MSG_DEBUG, "AUTH: %s (vlan_id=%d alg=%d idx=%d key_len=%d)",
+		   __func__, vlan_id, alg, idx, (int) key_len);
+	if (addr)
+		wpa_printf(MSG_DEBUG, "AUTH: addr=" MACSTR, MAC2STR(addr));
+
+	if (!key_len || !addr)
+		return 0;
+
+	os_memcpy(key_param.mac_addr, addr, ETH_ALEN);
+	key_param.idx =  idx;
+	os_memcpy(key_param.key_data, key, key_len);
+	key_param.key_len = key_len;
+
+	os_memset(send_buffer, 0, 64);
+	send_buffer[0] = EVENT_SETKEY;
+	WPA_PUT_BE16(send_buffer+1, sizeof(key_param));
+	os_memcpy(send_buffer+3, &key_param, sizeof(key_param));
+
+	wpa_hexdump(MSG_ERROR, "SUPP: set_key - key", key, key_len);
+
+	send(eapol_1905_sock_tx, send_buffer, 64, 0);
+
+	if (!is_broadcast_ether_addr(addr)) {
+		int i = 0;
+		for (i = 0; i < wpa->sta_count; ) {
+			if (wpa->auth[i]) {
+				wpa_printf(MSG_DEBUG, "AUTH: i=%d addr= " MACSTR, i, MAC2STR(addr));
+				if (!os_memcmp(wpa->supp_addr[i], addr, ETH_ALEN)) {
+					wpa_auth_sta_deinit(wpa->auth[i]);
+					wpa->sta_count--;
+					break;
+				}
+				i++;
+			}
+		}
+	}
+	return 0;
+}
+
+static int wpa_1905_send_eapol(void *ctx, const u8 *addr, const u8 *data,
+			   size_t data_len, int encrypt)
+{
+	struct wpa_1905 *wpa = ctx;
+	u8 send_buffer[1024] = {0} ;
+	os_memset(send_buffer, 0, 1024);
+	send_buffer[0] = EVENT_1905_EAPOL;
+	WPA_PUT_BE16(send_buffer + 1 , data_len + ETH_ALEN);
+
+	wpa_printf(MSG_DEBUG, "AUTH: TX EAPOL addr=" MACSTR,
+		   MAC2STR(addr));
+	os_memcpy(send_buffer+3, addr, ETH_ALEN);
+	os_memcpy(send_buffer+9, data, data_len);
+	send(eapol_1905_sock_tx, send_buffer, 1024, 0);
+}
+
+
+static int assoc_init(struct wpa_1905 *wpa)
+{
+	int i = wpa->sta_count ;
+	wpa->auth[i] = wpa_auth_sta_init(wpa->auth_group,
+					 wpa->supp_addr[i], NULL);
+	if (!wpa->auth[i]) {
+		wpa_printf(MSG_DEBUG, "AUTH: wpa_auth_sta_init() failed");
+		return -1;
+	}
+
+	if (wpa_auth_pmksa_add2(wpa->auth_group, wpa->supp_addr[i], wpa->psk,
+				PMK_LEN, wpa->pmkid, 0,
+				conf.wpa_key_mgmt) < 0) {
+		wpa_printf(MSG_ERROR, "DPP: Failed to add PMKSA cache entry");
+		return -1;
+	}
+
+	if (wpa_validate_wpa_ie(wpa->auth_group, wpa->auth[i], 2412,
+				wpa->supp_ie, wpa->supp_ie_len, NULL, 0,
+				NULL, 0, NULL, 0) != WPA_IE_OK) {
+		wpa_printf(MSG_DEBUG, "AUTH: wpa_validate_wpa_ie() failed");
+		wpa_auth_sta_deinit(wpa->auth[i]);
+		wpa->sta_count--;
+		return -1;
+	}
+
+	wpa_auth_sm_event(wpa->auth[i], WPA_ASSOC);
+
+	wpa_auth_sta_associated(wpa->auth_group, wpa->auth[i]);
+
+	return 0;
+}
+
+static int ap_init(struct wpa_1905 *wpa, struct sec_config *sec_conf)
+{
+
+	wpa_printf(MSG_DEBUG, "AUTH: Initializing group state machine");
+
+	os_memcpy(wpa->auth_addr, sec_conf->auth_addr, ETH_ALEN);
+	os_memcpy(wpa->psk, sec_conf->psk, PMK_LEN);
+
+	os_memset(&conf, 0, sizeof(conf));
+	conf.wpa = sec_conf->wpa_proto;
+	conf.wpa_key_mgmt = sec_conf->key_mgmt;
+	conf.wpa_pairwise = sec_conf->pairwise_cipher;
+	conf.rsn_pairwise = sec_conf->pairwise_cipher;
+	conf.wpa_group = sec_conf->group_cipher;
+	conf.ieee80211w = sec_conf->ieee80211w;
+	conf.group_mgmt_cipher = sec_conf->mgmt_group_cipher;
+	conf.no_group_delay = 1;
+
+	conf.eapol_version = 2;
+	conf.wpa_group_update_count = 4;
+	conf.wpa_pairwise_update_count = 4;
+
+	wpa->auth_cb.logger = auth_logger;
+	wpa->auth_cb.send_eapol = wpa_1905_send_eapol;
+	wpa->auth_cb.get_psk = auth_get_psk;
+	wpa->auth_cb.set_key = auth_set_key,
+
+
+	wpa->auth_group = wpa_init(wpa->auth_addr, &conf, &wpa->auth_cb, wpa);
+
+
+	wpa->sta_count = 0;
+	if (!wpa->auth_group) {
+		wpa_printf(MSG_DEBUG, "AUTH: wpa_init() failed");
+		return -1;
+	}
+
+	return 0;
+}
+
+static void eapol_1905_rx_receive(int sock, void *eloop_ctx, void *sock_ctx)
+{
+	u8 buf[4096];
+	int res;
+	struct sockaddr_storage from;
+	socklen_t fromlen = sizeof(from);
+	const int reply_size = 4096;
+	int reply_len;
+	int level = MSG_DEBUG;
+	struct wpa_1905 *wpa = eloop_ctx;
+
+	res = recvfrom(sock, buf, sizeof(buf) - 1, 0,
+		       (struct sockaddr *) &from, &fromlen);
+	if (res < 0) {
+		wpa_printf(MSG_ERROR, "recvfrom(ctrl_iface): %s",
+			   strerror(errno));
+		return;
+	}
+
+	buf[res] = '\0';
+
+	wpa_printf(MSG_ERROR, "Event:%d", buf[0]);
+
+	if (buf[0] == EVENT_1905_ASSOC) {
+		struct assoc *new_assoc = (struct assoc *) &buf[1];
+
+		if (wpa->sta_count < MAX_AUTH) {
+			os_memcpy(wpa->supp_addr[wpa->sta_count],
+				  new_assoc->mac_addr, ETH_ALEN);
+			wpa->supp_ie_len = new_assoc->supp_ie_len;
+			os_memcpy(wpa->supp_ie,
+				  new_assoc->supp_ie, wpa->supp_ie_len);
+			os_memcpy(wpa->pmkid, new_assoc->pmkid, PMKID_LEN);
+
+			if (!assoc_init(wpa))
+				wpa->sta_count++;
+		} else {
+			wpa_printf(MSG_ERROR, "STA Assoc max count reached");
+		}
+	} else if (buf[0] == EVENT_1905_EAPOL) {
+		int i = 0;
+/* Compare with src addr */
+
+		for (i = 0; i < wpa->sta_count; i++) {
+			if (!os_memcmp(wpa->supp_addr[i], buf + 3,  ETH_ALEN))
+				break;
+		}
+		wpa_printf(MSG_DEBUG, "AUTH: RX EAPOL from i=%d addr=" MACSTR,
+			   i, MAC2STR(wpa->supp_addr[i]));
+
+		if (i == wpa->sta_count)
+			return ;
+
+		wpa_receive(wpa->auth_group, wpa->auth[i], buf + 9,
+			    WPA_GET_BE16(&buf[1]) - ETH_ALEN);
+	} else if (buf[0] == EVENT_AP_SEC_CONFIG) {
+		int i = 0;
+		if (wpa->sta_count > 0) {
+			for (i = 0; i < wpa->sta_count; ) {
+				if (wpa->auth[i]) {
+					i++;
+					wpa_auth_sta_deinit(wpa->auth[i]);
+				}
+			}
+			wpa->sta_count = 0;
+		}
+		if (wpa->auth_group)
+			wpa_deinit(wpa->auth_group);
+
+		ap_init(wpa, (struct sec_config *) &buf[1]);
+	} else if (buf[0] == EVENT_GROUP_KEY) {
+		if (wpa->auth_group)
+			wpa_auth_reconfig_group_keys(wpa->auth_group);
+	}
+
+}
+
+int eapol_1905_init()
+{
+	struct sockaddr_in address;
+
+	eapol_1905_sock = socket(AF_INET, SOCK_DGRAM, 0);
+	if (eapol_1905_sock < 0) {
+		wpa_printf(MSG_ERROR, "nl80211: socket(PF_PACKET, SOCK_DGRAM, 1905 ETH_P_PAE) failed: %s",
+			   strerror(errno));
+		return -1 ;
+	}
+
+	memset((char *)&address, 0, sizeof(address));
+	address.sin_family = AF_INET;
+	address.sin_addr.s_addr = htonl((127 << 24) | 1);
+	address.sin_port = htons(RX_PORT);
+
+	if (bind(eapol_1905_sock, (struct sockaddr *)&address, sizeof(address)) < 0) {
+		wpa_printf(MSG_ERROR, "nl80211: socket(PF_PACKET, SOCK_DGRAM, 1905 ETH_P_PAE) failed: %s",
+			   strerror(errno));
+		close(eapol_1905_sock);
+		return -1;
+	}
+
+	eapol_1905_sock_tx = socket(AF_INET, SOCK_DGRAM, 0);
+	if (eapol_1905_sock_tx < 0) {
+		wpa_printf(MSG_ERROR, "nl80211: socket(PF_PACKET, SOCK_DGRAM, 1905 ETH_P_PAE) failed: %s",
+			   strerror(errno));
+		return -1 ;
+	}
+	address.sin_port = htons(TX_PORT);
+
+	if (connect(eapol_1905_sock_tx, (struct sockaddr *)&address, sizeof(address)) < 0) {
+		wpa_printf(MSG_ERROR, " connect failed :%s ", strerror(errno));
+		close(eapol_1905_sock_tx);
+		return -1;
+	}
+
+	os_memset(&wpa, 0, sizeof(wpa));
+	if (eloop_register_read_sock(eapol_1905_sock, eapol_1905_rx_receive, &wpa, NULL)) {
+		close(eapol_1905_sock);
+		wpa_printf(MSG_INFO, "nl80211: Could not register read socket for 1905 eapol");
+		return -1;
+	}
+
+	return 0;
+}
+
+void eapol_1905_deinit(void *priv)
+{
+	int i = 0;
+	wpa_printf(MSG_DEBUG, "EAPOL 1905_deinit");
+
+	if (eapol_1905_sock_tx > 0) {
+		close(eapol_1905_sock_tx);
+		eapol_1905_sock_tx = -1;
+	}
+
+	if (eapol_1905_sock > -1) {
+		eloop_unregister_read_sock(eapol_1905_sock);
+		close(eapol_1905_sock);
+		eapol_1905_sock = -1;
+	}
+
+	if (wpa.sta_count > 0) {
+		for (i = 0; i < wpa.sta_count; ) {
+			if (wpa.auth[i]) {
+				i++;
+				wpa_auth_sta_deinit(wpa.auth[i]);
+			}
+		}
+		wpa.sta_count = 0;
+	}
+
+	if (wpa.auth_group)
+		wpa_deinit(wpa.auth_group);
+	return;
+
+}
diff --git a/src/drivers/eap1905.h b/src/drivers/eap1905.h
new file mode 100644
index 0000000000..8593f0da33
--- /dev/null
+++ b/src/drivers/eap1905.h
@@ -0,0 +1,7 @@
+/*
+ * Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+ * SPDX-License-Identifier: ISC
+ */
+
+int eapol_1905_init(void);
+void eapol_1905_deinit(void);
diff --git a/src/drivers/eap1905_sta.c b/src/drivers/eap1905_sta.c
new file mode 100644
index 0000000000..521bfa09ed
--- /dev/null
+++ b/src/drivers/eap1905_sta.c
@@ -0,0 +1,445 @@
+
+/*
+ * Testing tool for EAPOL-Key Supplicant/Authenticator routines
+ * Copyright (c) 2006-2019, Jouni Malinen <j@w1.fi>
+ *
+ * This software may be distributed under the terms of the BSD license.
+ * See README for more details.
+ */
+
+
+#include "utils/includes.h"
+
+#include "utils/common.h"
+#include "utils/eloop.h"
+#include "rsn_supp/wpa.h"
+#include "rsn_supp/pmksa_cache.h"
+
+#define RX_PORT 9902
+#define TX_PORT 9903
+
+struct wpa_1905 {
+	int no_callback;
+	u8 auth_addr[ETH_ALEN];
+	u8 sta_supp_addr[ETH_ALEN];
+	struct wpa_sm *supp;
+	struct wpa_sm_ctx *ctx;
+
+	u8 supp_ie[80];
+	size_t supp_ie_len;
+	u8 psk[PMK_LEN];
+	u8 pmkid[PMKID_LEN];
+
+};
+
+struct sta_sec_config {
+	u8 auth_addr[ETH_ALEN];
+	u8 supp_addr[ETH_ALEN];
+	u8 psk[PMK_LEN];
+	int pairwise_cipher;
+	int group_cipher;
+	int key_mgmt;
+	int wpa_proto;
+	int mgmt_group_cipher;
+	int ieee80211w;
+	size_t supp_ie_len;
+	u8 supp_ie[80];
+	u8 pmkid[PMKID_LEN];
+} STRUCT_PACKED;
+
+struct key_params {
+	u8 mac_addr[ETH_ALEN];
+	int idx;
+	u8 key_data[32];
+	size_t key_len;
+} STRUCT_PACKED;
+
+enum events {
+	EVENT_AP_SEC_CONFIG,
+	EVENT_STA_SEC_CONFIG,
+	EVENT_1905_ASSOC,
+	EVENT_DISASSOC,
+	EVENT_1905_EAPOL,
+	EVENT_SETKEY,
+};
+
+int eapol_1905_sock = -1;
+int eapol_1905_sock_tx = -1;
+static struct wpa_1905 wpa;
+
+static int  supp_get_beacon_ie(void *ctx)
+{
+	struct wpa_1905 *wpa = ctx;
+	const u8 *ie;
+
+/* Beacon rsn IE must come from MAP */
+
+	static const u8 rsne[] = {
+		0x30, 0x14, 0x01, 0x00, 0x00, 0x0f, 0xac, 0x0a,
+		0x01, 0x00, 0x00, 0x0f, 0xac, 0x0a, 0x01, 0x00,
+		0x50, 0x6f, 0x9a, 0x02, 0x00, 0x00
+	};
+
+	wpa_printf(MSG_ERROR, "SUPP: %s", __func__);
+
+	ie = rsne;
+	if (ie[0] == WLAN_EID_RSN)
+		return wpa_sm_set_ap_rsn_ie(wpa->supp, ie, 2 + ie[1]);
+	return wpa_sm_set_ap_wpa_ie(wpa->supp, ie, 2 + ie[1]);
+}
+
+static int supp_get_bssid(void *ctx, u8 *bssid)
+{
+	struct wpa_1905 *wpa = ctx;
+	wpa_printf(MSG_ERROR, "SUPP: %s", __func__);
+	os_memcpy(bssid, wpa->auth_addr, ETH_ALEN);
+	return 0;
+}
+
+static void supp_set_state(void *ctx, enum wpa_states state)
+{
+	wpa_printf(MSG_ERROR, "SUPP: %s(state=%d)", __func__, state);
+}
+
+static int supp_ether_send(void *ctx, const u8 *dest, u16 proto, const u8 *buf,
+			   size_t len)
+{
+	struct wpa_1905 *wpa = ctx;
+	int i;
+
+	u8 send_buffer[1024] = {0} ;
+	send_buffer[0] = EVENT_1905_EAPOL;
+	wpa_printf(MSG_DEBUG, "TX EPAOL : data_len:%d ", len);
+
+	wpa_printf(MSG_DEBUG, "AUTH: TX EAPOL dest addr=" MACSTR,
+		   MAC2STR(dest));
+
+	WPA_PUT_BE16(send_buffer + 1, len + ETH_ALEN);
+	os_memcpy(send_buffer+3, wpa->sta_supp_addr, ETH_ALEN);
+	os_memcpy(send_buffer+9, buf, len);
+
+	send(eapol_1905_sock_tx, send_buffer , 1024 , 0);
+	wpa_printf(MSG_ERROR, "SUPP: %s(dest=" MACSTR " proto=0x%04x len=%lu)",
+		   __func__, MAC2STR(dest), proto, (unsigned long) len);
+	/* Key info */
+	if (WPA_GET_BE16(buf+sizeof(struct ieee802_1x_hdr)+1) & WPA_KEY_INFO_INSTALL)
+		return 0;
+
+	return 0;
+}
+
+static u8 * supp_alloc_eapol(void *ctx, u8 type, const void *data,
+			     u16 data_len, size_t *msg_len, void **data_pos)
+{
+	struct ieee802_1x_hdr *hdr;
+	wpa_printf(MSG_ERROR, "SUPP: %s(type=%d data_len=%d)",
+		   __func__, type, data_len);
+
+	*msg_len = sizeof(*hdr) + data_len;
+	hdr = os_malloc(*msg_len);
+	if (hdr == NULL)
+		return NULL;
+
+	hdr->version = 2;
+	hdr->type = type;
+	hdr->length = host_to_be16(data_len);
+
+	if (data)
+		os_memcpy(hdr + 1, data, data_len);
+	else
+		os_memset(hdr + 1, 0, data_len);
+
+	if (data_pos)
+		*data_pos = hdr + 1;
+
+	return (u8 *) hdr;
+}
+
+
+static int supp_set_key(void *ctx, enum wpa_alg alg,
+			const u8 *addr, int key_idx, int set_tx,
+			const u8 *seq, size_t seq_len,
+			const u8 *key, size_t key_len,  enum key_flag key_flag)
+{
+	char send_buffer[64] = {0} ;
+	struct key_params key_param;
+	os_memset(&key_param, 0, sizeof(key_param));
+
+	wpa_printf(MSG_ERROR, "SUPP: %s(alg=%d addr=" MACSTR " key_idx=%d set_tx=%d)",
+		   __func__, alg, MAC2STR(addr), key_idx, set_tx);
+	wpa_hexdump(MSG_ERROR, "SUPP: set_key - seq", seq, seq_len);
+	wpa_hexdump(MSG_ERROR, "SUPP: set_key - key", key, key_len);
+
+	if (addr)
+		wpa_printf(MSG_DEBUG, "SUPP: addr=" MACSTR, MAC2STR(addr));
+	if (!key_len)
+		return 0;
+
+	os_memcpy(key_param.mac_addr, addr, ETH_ALEN);
+	key_param.idx =  key_idx;
+	os_memcpy(key_param.key_data, key, key_len);
+	key_param.key_len = key_len;
+
+	os_memset(send_buffer, 0, 64);
+	send_buffer[0] = EVENT_SETKEY;
+	WPA_PUT_BE16(send_buffer+1, sizeof(key_param));
+	os_memcpy(send_buffer+3, &key_param, sizeof(key_param));
+
+	wpa_hexdump(MSG_ERROR, "SUPP: set_key - key", key, key_len);
+
+	send(eapol_1905_sock_tx, send_buffer, 64, 0);
+
+	return 0;
+}
+
+
+static int supp_mlme_setprotection(void *ctx, const u8 *addr,
+				   int protection_type, int key_type)
+{
+	wpa_printf(MSG_ERROR, "SUPP: %s(addr=" MACSTR " protection_type=%d "
+		   "key_type=%d)",
+		   __func__, MAC2STR(addr), protection_type, key_type);
+
+	if (protection_type == 3) {
+		wpa_sm_notify_assoc(wpa.supp, wpa.auth_addr);
+		wpa_printf(MSG_ERROR, "SUPP: Reinit assoc ");
+	}
+	return 0;
+}
+
+
+static void supp_cancel_auth_timeout(void *ctx)
+{
+	wpa_printf(MSG_ERROR, "SUPP: %s", __func__);
+}
+
+
+static void * supp_get_network_ctx(void *ctx)
+{
+	return (void *) 1;
+}
+
+
+static void supp_deauthenticate(void *ctx, u16 reason_code)
+{
+	wpa_printf(MSG_ERROR, "SUPP: %s(%d)", __func__, reason_code);
+}
+
+static enum wpa_states supp_get_state(void *ctx)
+{
+	return WPA_COMPLETED;
+}
+
+static int supp_add_pmkid(void *wpa_s, void *network_ctx,
+			  const u8 *bssid, const u8 *pmkid,
+			  const u8 *fils_cache_id,
+			  const u8 *pmk, size_t pmk_len,
+			  u32 pmk_lifetime, u8 pmk_reauth_threshold,
+			  int akmp)
+
+{
+	wpa_printf(MSG_ERROR, "SUPP: Add PMKID ");
+	return 0;
+}
+
+void wpa_sm_inits(struct wpa_sm_ctx *ctx)
+{
+	printf("%p\n", ctx->set_rekey_offload);
+}
+
+void supp_rekey_offload(void *ctx, const u8 *kek, size_t kek_len,
+			const u8 *kck, size_t kck_len,
+			const u8 *replay_ctr)
+{
+	printf("**************\n");
+	return;
+}
+
+static int sta_init(struct wpa_1905 *wpa, struct sta_sec_config *sec_conf)
+{
+
+	wpa->ctx = os_zalloc(sizeof(struct wpa_sm_ctx));
+	wpa->ctx->ctx = wpa;
+	wpa->ctx->msg_ctx = wpa;
+	wpa->ctx->set_state = supp_set_state;
+	wpa->ctx->get_bssid = supp_get_bssid;
+	wpa->ctx->ether_send = supp_ether_send;
+	wpa->ctx->get_beacon_ie = supp_get_beacon_ie;
+	wpa->ctx->alloc_eapol = supp_alloc_eapol;
+	wpa->ctx->set_key = supp_set_key;
+	wpa->ctx->mlme_setprotection = supp_mlme_setprotection;
+	wpa->ctx->cancel_auth_timeout = supp_cancel_auth_timeout;
+	wpa->ctx->get_network_ctx = supp_get_network_ctx;
+	wpa->ctx->deauthenticate = supp_deauthenticate;
+	wpa->ctx->get_state = supp_get_state;
+	wpa->ctx->set_rekey_offload = NULL;
+	wpa->ctx->add_pmkid = supp_add_pmkid;
+
+	wpa->no_callback = 1 ;
+
+	os_memcpy(wpa->auth_addr, sec_conf->auth_addr, ETH_ALEN);
+	os_memcpy(wpa->sta_supp_addr, sec_conf->supp_addr, ETH_ALEN);
+	os_memcpy(wpa->psk, sec_conf->psk, PMK_LEN);
+	os_memcpy(wpa->pmkid, sec_conf->pmkid, PMKID_LEN);
+	os_memcpy(wpa->supp_ie, sec_conf->supp_ie, sec_conf->supp_ie_len);
+	wpa->supp_ie_len = sec_conf->supp_ie_len ;
+
+	wpa->supp = wpa_sm_init(wpa->ctx);
+
+	wpa_sm_set_pmk(wpa->supp, wpa->psk, PMK_LEN, NULL, NULL);
+
+	if (!wpa->supp) {
+		wpa_printf(MSG_ERROR, "SUPP: wpa_sm_init() failed");
+		return -1;
+	}
+
+	wpa_sm_set_own_addr(wpa->supp, wpa->sta_supp_addr);
+	{
+		wpa_sm_set_param(wpa->supp, WPA_PARAM_RSN_ENABLED, 1);
+		wpa_sm_set_param(wpa->supp, WPA_PARAM_PROTO, sec_conf->wpa_proto);
+		wpa_sm_set_param(wpa->supp, WPA_PARAM_PAIRWISE,
+				 sec_conf->pairwise_cipher);
+		wpa_sm_set_param(wpa->supp, WPA_PARAM_GROUP, sec_conf->group_cipher);
+		wpa_sm_set_param(wpa->supp, WPA_PARAM_KEY_MGMT,
+				 sec_conf->key_mgmt);
+		wpa_sm_set_param(wpa->supp, WPA_PARAM_MFP,
+				 sec_conf->ieee80211w);
+	}
+
+	wpa_sm_pmksa_cache_add(wpa->supp, wpa->psk, PMK_LEN,  wpa->pmkid, wpa->auth_addr, NULL);
+
+	if (wpa_sm_set_assoc_wpa_ie_default(wpa->supp, wpa->supp_ie,
+					    &wpa->supp_ie_len) < 0) {
+		wpa_printf(MSG_ERROR, "SUPP: wpa_sm_set_assoc_wpa_ie_default()"
+			   " failed");
+		return -1;
+	}
+
+	wpa_sm_notify_assoc(wpa->supp, wpa->auth_addr);
+
+	return 0;
+}
+
+static void eapol_1905_rx_receive(int sock, void *eloop_ctx, void *sock_ctx)
+{
+	u8 buf[4096];
+	int res;
+	struct sockaddr_storage from;
+	socklen_t fromlen = sizeof(from);
+	const int reply_size = 4096;
+	int reply_len;
+	int level = MSG_DEBUG;
+	struct wpa_1905 *wpa = eloop_ctx;
+
+	wpa_printf(MSG_ERROR, "eapol_1905_rx_receive called ..");
+
+	res = recvfrom(sock, buf, sizeof(buf) - 1, 0,
+			(struct sockaddr *) &from, &fromlen);
+	if (res < 0) {
+		wpa_printf(MSG_ERROR, "recvfrom(ctrl_iface): %s",
+			   strerror(errno));
+		return;
+	}
+	buf[res] = '\0';
+
+	wpa_printf(MSG_ERROR, "Event:%d ", buf[0]);
+
+/* The first byte is type of info EAPOL / ASSOC. second byte is len of EAPOL or assoc info */
+
+	if (buf[0] == EVENT_1905_EAPOL) {
+		int len = 0;
+		char *msg;
+
+		if (!os_memcmp(wpa->sta_supp_addr, buf+9,  ETH_ALEN))
+			wpa_printf(MSG_ERROR, "eapol_len:%d ", WPA_GET_BE16(&buf[1])-ETH_ALEN);
+
+		wpa_printf(MSG_ERROR, "eapol_len:%d ", WPA_GET_BE16(&buf[1]) - ETH_ALEN);
+
+		len = WPA_GET_BE16(&buf[1]) - ETH_ALEN ;
+		msg = os_malloc(len);
+		os_memcpy(msg, buf+9, len);
+
+		wpa_sm_rx_eapol(wpa->supp, wpa->auth_addr, msg, len);
+		wpa_printf(MSG_ERROR, "wpa_sm_rx_eapol done ");
+		os_free(msg);
+
+	} else if (buf[0] == EVENT_STA_SEC_CONFIG) {
+		if (wpa->supp)
+			wpa_sm_deinit(wpa->supp);
+		sta_init(wpa, (struct sta_sec_config *) &buf[1]);
+	} else if (buf[0] == EVENT_1905_ASSOC) {
+		if (wpa->supp)
+			wpa_sm_notify_assoc(wpa->supp, wpa->auth_addr);
+	}
+
+}
+
+
+
+
+int eapol_1905_init()
+{
+	struct sockaddr_in address;
+
+	wpa_printf(MSG_INFO, "nl80211: eapol_1905_init");
+
+	eapol_1905_sock = socket(AF_INET, SOCK_DGRAM, 0);
+	if (eapol_1905_sock < 0) {
+		wpa_printf(MSG_ERROR, "nl80211: socket(PF_PACKET, SOCK_DGRAM, 1905 ETH_P_PAE) failed: %s",
+			   strerror(errno));
+		return -1 ;
+	}
+
+	memset((char *)&address, 0, sizeof(address));
+	address.sin_family = AF_INET;
+	address.sin_addr.s_addr = htonl((127 << 24) | 1);
+	address.sin_port = htons(RX_PORT);
+
+	if (bind(eapol_1905_sock, (struct sockaddr *)&address, sizeof(address)) < 0) {
+		wpa_printf(MSG_ERROR, "nl80211: socket(PF_PACKET, SOCK_DGRAM, 1905 ETH_P_PAE) failed: %s",
+			   strerror(errno));
+		close(eapol_1905_sock);
+		return -1;
+	}
+
+	eapol_1905_sock_tx = socket(AF_INET, SOCK_DGRAM, 0);
+	if (eapol_1905_sock_tx < 0) {
+		wpa_printf(MSG_ERROR, "nl80211: socket(PF_PACKET, SOCK_DGRAM, 1905 ETH_P_PAE) failed: %s",
+			   strerror(errno));
+		return -1 ;
+	}
+
+	address.sin_port = htons(TX_PORT);
+
+	if (connect(eapol_1905_sock_tx, (struct sockaddr *)&address, sizeof(address)) < 0) {
+		wpa_printf(MSG_ERROR, " connect failed :%s ", strerror(errno));
+		close(eapol_1905_sock_tx);
+		return -1;
+	}
+
+	os_memset(&wpa, 0, sizeof(wpa));
+	if (eloop_register_read_sock(eapol_1905_sock, eapol_1905_rx_receive, &wpa, NULL)) {
+		close(eapol_1905_sock);
+		wpa_printf(MSG_INFO, "nl80211: Could not register read socket for 1905 eapol");
+		return -1;
+	}
+
+	wpa_printf(MSG_INFO, "nl80211: eapol_1905_init done");
+	return 0;
+}
+
+void eapol_1905_deinit(void *priv)
+{
+	wpa_printf(MSG_DEBUG, "EAPOL 1905_deinit");
+	if (eapol_1905_sock_tx > 0) {
+		close(eapol_1905_sock_tx);
+		eapol_1905_sock_tx = -1;
+	}
+	if (eapol_1905_sock > -1) {
+		eloop_unregister_read_sock(eapol_1905_sock);
+		close(eapol_1905_sock);
+		eapol_1905_sock = -1;
+	}
+	wpa_sm_deinit(wpa.supp);
+	return;
+}
diff --git a/wpa_supplicant/Makefile b/wpa_supplicant/Makefile
index 83a28de67b..d219986120 100644
--- a/wpa_supplicant/Makefile
+++ b/wpa_supplicant/Makefile
@@ -1936,6 +1936,7 @@ OBJS += ../src/drivers/driver_common.o
 OBJS_priv += ../src/drivers/driver_common.o
 
 OBJS += wpa_supplicant.o events.o bssid_ignore.o wpas_glue.o scan.o
+OBJS += ../src/drivers/eap1905_sta.o
 OBJS_t := $(OBJS) $(OBJS_l2) eapol_test.o
 OBJS_t += ../src/radius/radius_client.o
 OBJS_t += ../src/radius/radius.o
diff --git a/wpa_supplicant/ctrl_iface_unix.c b/wpa_supplicant/ctrl_iface_unix.c
index 2052873d57..968560a067 100644
--- a/wpa_supplicant/ctrl_iface_unix.c
+++ b/wpa_supplicant/ctrl_iface_unix.c
@@ -436,6 +436,9 @@ static void wpa_supplicant_ctrl_iface_msg_cb(void *ctx, int level,
 	if (wpa_s == NULL)
 		return;
 
+	if (wpa_s->no_callback)
+		return;
+
 	gpriv = wpa_s->global->ctrl_iface;
 
 	if (type != WPA_MSG_NO_GLOBAL && gpriv &&
diff --git a/wpa_supplicant/main.c b/wpa_supplicant/main.c
index 9229eb51f3..fa606dd65a 100644
--- a/wpa_supplicant/main.c
+++ b/wpa_supplicant/main.c
@@ -17,6 +17,7 @@
 #include "wpa_supplicant_i.h"
 #include "driver_i.h"
 #include "p2p_supplicant.h"
+#include "drivers/eap1905.h"
 
 
 static void usage(void)
@@ -389,9 +390,12 @@ int main(int argc, char *argv[])
 		exitcode = wpa_supplicant_init_match(global);
 #endif /* CONFIG_MATCH_IFACE */
 
-	if (exitcode == 0)
+	if (exitcode == 0) {
+		eapol_1905_init();
 		exitcode = wpa_supplicant_run(global);
+	}
 
+	eapol_1905_deinit();
 	wpa_supplicant_deinit(global);
 
 	fst_global_deinit();
diff --git a/wpa_supplicant/wpa_supplicant_i.h b/wpa_supplicant/wpa_supplicant_i.h
index 56f100a447..8a48a86d4f 100644
--- a/wpa_supplicant/wpa_supplicant_i.h
+++ b/wpa_supplicant/wpa_supplicant_i.h
@@ -678,6 +678,7 @@ struct ml_sta_link_info {
  * core functions.
  */
 struct wpa_supplicant {
+	int no_callback;
 	struct wpa_global *global;
 	struct wpa_radio *radio; /* shared radio context */
 	struct dl_list radio_list; /* list head: struct wpa_radio::ifaces */
-- 
2.34.1

